src/waltham/waltham-connection.c

```diff
@@ -118,6 +119,62 @@ static const struct wth_display_listener display_listener = {
 
 /* END wthp_display client implementation */
 
+/* BEGIN wthp_display server implementation */
+
+/* TODO: these three declarations are copied from waltham-server.h, which can’t
+ * be included together with waltham-client.h.  Find a way to make it possible
+ * to include them both, or split this server implementation in another file.
+ */
+
+void
+wthp_callback_send_done (struct wthp_callback * wthp_callback, uint32_t callback_data);
+
+struct wth_display_interface {
+	void (*client_version) (struct wth_display * wth_display, uint32_t client_version);
+	void (*sync) (struct wth_display * wth_display, struct wthp_callback * callback);
+	void (*get_registry) (struct wth_display * wth_display, struct wthp_registry * registry);
+};
+
+static inline void
+wth_display_set_interface(struct wth_display *self, const struct wth_display_interface *funcs, void *user_data)
+{
+	wth_object_set_listener((struct wth_object *)self, (void (**)(void)) funcs, user_data);
+}
+
+static void
+display_handle_client_version(struct wth_display *wth_display,
+                              uint32_t client_version)
+{
+	wth_object_post_error((struct wth_object *)wth_display, 0,
+	                      "unimplemented: %s", __func__);
+}
+
+static void
+display_handle_sync(struct wth_display *wth_display,
+                    struct wthp_callback *callback)
+{
+	wth_debug("Client requested wth_display.sync");
+	wthp_callback_send_done(callback, 0);
+	wthp_callback_free(callback);
+}
+
+static void
+display_handle_get_registry(struct wth_display *wth_display,
+                            struct wthp_registry *registry)
+{
+	struct wth_object *disp_object = (struct wth_object *)wth_display;
+	struct wth_connection *conn = disp_object->connection;
+	conn->registry_callback(registry, conn->registry_callback_user_data);
```

Comment:
I think here if `registry_callback` is NULL, we should call `wth_abort()` with a suitable error message saying the server implementation in the libwaltham user is buggy.

Other than that, looks good!